src/FileCache.js

```diff
@@ -0,0 +1,221 @@
+// Copyright (c) 2017 Electric Imp
+// This file is licensed under the MIT License
+// http://opensource.org/licenses/MIT
+
+'use strict';
+
+const fs = require('fs-extra');
+const path = require('path');
+const minimatch = require('minimatch');
+const md5 = require('md5');
+const HttpReader = require('./Readers/HttpReader');
+const GithubReader = require('./Readers/GithubReader');
+
+const DEFAULT_EXCLUDE_FILE_NAME = 'builder-cache.exclude';
+const CACHED_READERS = [GithubReader, HttpReader];
+const CACHE_LIFETIME = 1; // in days
+
+class FileCache {
+
+  constructor(machine) {
+    this._useCache = false;
+    this._cacheDir = '.' + path.sep + '.builder-cache';
+    this._excludeList = [];
+    this._machine = machine;
+    this._outdateTime = CACHE_LIFETIME * 86400000; // precalc milliseconds in one day
+  }
+
+  /**
+   * Transform url or github link to path and filename
+   * It is important, that path and filename are unique,
+   * because collision can break the build
+   * @param {string} link link to the file
+   * @return {string} folder and name, where cache file can be found
+   * @private
+   */
+  _getCachedPath(link) {
+    link = link.replace(/^github\:/, 'github#'); // replace ':' for '#' in github protocol
+    link = link.replace(/\:\/\//, '#'); // repalce '://' for '#' in url
+    link = link.replace(/\./g, '_');
+    link = link.replace(/\//g, '-'); // replace '/' for '-'
+    link = link.replace(/\?(.*)/g, ''); // delete get parameters from url
+    if (link.length > 250) {
+      const startPart = link.substr(0, 100);
+      const endPart = link.substr(link.length - 100);
+      const middlePart = md5(link);
```

Comment:
I would give https://cyan4973.github.io/xxHash/ a try instead (to speed up the compilation process).